### PR TITLE
[trusted-types] Update definitions to match Trusted Types specification

### PR DESCRIPTION
**Pull request by bjarkler**
This addresses divergence in the current type definitions from the
Trusted Types specification. The types were also rearranged to match
spec order to make comparison easier. Tests were also updated
accordingly.

Please fill in this template.

- [X] Use a meaningful title for the pull request. Include the name of the package modified.
- [X] Test the change in your own code. (Compile and run.)
- [X] Add or edit tests to reflect the change. (Run with `npm test`.)
- [X] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [X] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [X] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [X] Provide a URL to documentation or source code which provides context for the suggested changes: https://w3c.github.io/webappsec-trusted-types/dist/spec/

### CODE REVIEW

File: types/trusted-types/index.d.ts
Comment (engelsdamien) [inline types/trusted-types/index.d.ts:62]: could you delete the uppercase `TrustedTypes` ?

**Review by engelsdamien — APPROVED** [review]

File: types/trusted-types/index.d.ts
Comment (bjarkler) [reply types/trusted-types/index.d.ts:60]: Done.

File: types/trusted-types/index.d.ts
Comment (engelsdamien) [inline types/trusted-types/index.d.ts:27]: we can actually also preserve the type of the `create*` functions to make this more typed:

```typescript
type TypedPolicy<Options extends TrustedTypePolicyOptions> = {
  name: string,
  createHTML(...args: Parameters<NonNullable<Options['createHTML']>>): TrustedHTML;
  createScript(...args: Parameters<NonNullable<Options['createScript']>>):  TrustedScript;
  createScriptURL(...args: Parameters<NonNullable<Options['createScriptURL']>>): TrustedScriptURL;
}

interface TrustedTypePolicyFactory {
  createPolicy<Options extends TrustedTypePolicyOptions>(
    policyName: string,
    policyOptions?: Options,
  ): Pick<TypedPolicy<Options>, 'name'|(keyof Options&keyof TrustedTypePolicyOptions)>;
}
```

File: types/trusted-types/index.d.ts
Comment (engelsdamien) [inline types/trusted-types/index.d.ts:31]: I think it would be a bit more ts idiomatic if all the `is*` functions took `unknown` instead of `any`

**Review by engelsdamien — CHANGES_REQUESTED** [review]

**Review by engelsdamien — APPROVED** [review]

File: types/trusted-types/index.d.ts
Comment (peterblazejewicz) [inline types/trusted-types/index.d.ts:63]:
```suggestion
    interface FullTypedPolicy<Options extends TrustedTypePolicyOptions> {
        name: string;
        createHTML(...args: Args<Options, 'createHTML'>): TrustedHTML;
        createScript(...args: Args<Options, 'createScript'>): TrustedScript;
        createScriptURL(...args: Args<Options, 'createScriptURL'>): TrustedScriptURL;
    }
```

File: types/trusted-types/tslint.json
Comment (peterblazejewicz) [inline types/trusted-types/tslint.json:6]:
```suggestion
{ "extends": "dtslint/dt.json" }
```
if you accept a comment about L.48 already showing interfaces are used to define methods, please accept that suggestion

**Review by peterblazejewicz — APPROVED** [review]
LGTM
The 3.1 will be the oldest tested on DT version in a month

**Comment by peterblazejewicz** [comment]
folks, quesiton, not related to the changes directly:
should users be able to use that module in global context? (taken from the docs:
```ts
if (window.trustedTypes && trustedTypes.createPolicy) {
    // Feature testing
    const escapeHTMLPolicy = trustedTypes.createPolicy('myEscapePolicy', {
        createHTML: string => string.replace(/\</g, '&lt;'),
    });
}

const p = trustedTypes.createPolicy('foo', {});
document.body.innerHTML = p.createHTML('foo'); // works
document.body.innerHTML = 'foo'; // but this one works too (no enforcement).
```

**Comment by engelsdamien** [comment]
technically yes, this is a browser API, so it would be available in the global context.

but IIRC we decided to make `trustedTypes` explicitly be only available on window & re-export it as a module so that it is easier to polyfill in node. 

@Siegrift has more context on this though

**Review by engelsdamien — APPROVED** [review]

File: types/trusted-types/index.d.ts
Comment (bjarkler) [reply types/trusted-types/index.d.ts]: Thanks, fixed. I originally did this to separate FullTypedPolicy from the rest of the interfaces since it is not part of the spec. But @engelsdamien pointed out to me that the correct approach is to move these helpers out of `declare global`, which I did.

File: types/trusted-types/tslint.json
Comment (bjarkler) [reply types/trusted-types/tslint.json]: Done.

**Review by peterblazejewicz — APPROVED** [review]

**Comment by Siegrift** [comment]
@peterblazejewicz @engelsdamien It's been a while, but IIRC the decision was based on a fact that some frameworks (e.g. next) has some code executed server side and simple check like:
```js
if (trustedTypes) {....}
```
would create Reference error.

**Comment by bjarkler** [comment]
I think this should be ready to merge? @typescript-bot :)

**Comment by engelsdamien** [comment]
Ready to merge

**Review by engelsdamien — APPROVED** [review]

**Comment by bjarkler** [comment]
@peterblazejewicz Do you know whether there is anything else that needs to be done before the PR can be merged? I'm wondering whether @typescript-bot is deadlocked here, especially since I added changes to a different package after I opened the PR, which the bot may not be used to.

**Review by peterblazejewicz — APPROVED** [review]

**Comment by peterblazejewicz** [comment]
@typescript-bot

**Comment by peterblazejewicz** [comment]
@bjarkler no idea here, I recall the bot being perpexled with some PR though. 
Can you try to `--amend` your last commit and push change (with different hash), so CI kicks in?
If this does not help, we will make a handle for Orta and create an issue here:
https://github.com/DefinitelyTyped/dt-mergebot/issues

**Review by peterblazejewicz — APPROVED** [review]
LGTM!

**Review by engelsdamien — APPROVED** [review]

**Comment by peterblazejewicz** [comment]
@orta @sandersn sorry folks for direct ping, thansk!
we've got somehow perplexed bot status here, PR is stuck in the needs review column:
https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/docs/admin.md#blessing-a-pr

**Comment by orta** [comment]
This one is on us (I've just improved the messaging in the bot for this state in https://github.com/DefinitelyTyped/dt-mergebot/pull/171 ) but basically the bot sees that this is editing 2 packages at once, and this requires an admin's influence in some form.

I'll happy just hit the merge button given that that is the intention - thanks for the ping, it was the right thing to do ( you can also nearly always get me instantly-ish via the discord https://discord.gg/typescript )

**Comment by elibarzilay** [comment]
@peterblazejewicz BTW, @orta's change would fix the misleading top comment in PR as happened here, but note that the PR was in "Needs Maintainer Review" because of multiple packages and this would still be the case.  (And I'm not sure how that can be improved.)

**Comment by bjarkler** [comment]
@elibarzilay I think this is something that should be made clear by the bot. When the PR has been reviewed and all checks have passed, a simple "Everything looks good. Since this PR affects multiple packages, please wait patiently for a DefinitelyTyped maintainer to do a final review." would be very helpful.

@orta Also a potential bug: typescript-bot published both dompurify and trusted-types to npm, but only added a comment here for dompurify.

**Comment by elibarzilay** [comment]
@bjarkler, it did that, only the line that was showing it:

> x Only a DT maintainer can approve changes **without tests**

was bogus.  The fix makes it show the proper line (as you can see right now in #46416, for example):

> x A DT maintainer needs to approve changes which affect more than one package

**Comment by peterblazejewicz** [comment]
thx!

**Pull request merged**
